### PR TITLE
fix: Improve lint error reporting

### LINKED ISSUE #8659: [Bug]-[2000]:Incorrect lint error for not using a comma separator after a function - shows up as missing semicolon
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When IIFE function is not terminated by a comma then lint error shows up as missing semicolon instead of missing comma and the position is incorrect - this causes a parse error

[LOOM DEMO](https://www.loom.com/share/4a518506db324936b987e38a5e7335f2)

### Steps To Reproduce

```markdown
1. Type out a JS function
2. Type out a second one below it without using a comma separator and observe the lint error and position
```

### Environment

Release

### Version

Cloud

### LINKED ISSUE #10825: [Feature]-[308]:Show correct lint error when async/await is used in a function
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

A user faced the following problem.

image

This is because, we are currently showing the following error message

image

Instead, we should tell the user to mark the parent function async.

### Why should this be worked on?

The current error message is not very helpful for the user to resolve the problem by himself/herself. We should provide an accurate error message.

### LINKED ISSUE #15099: [Bug]: Lint error not showing up for currentItem/currentRow when used in entities (Actions/Widgets/Queries) where it is not defined
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Lint error not showing up for currentItem/currentRow when used in entities (Actions/Widgets/Queries) where it is not defined
Screenshot 2022-07-10 at 10 41 09

### Steps To Reproduce

1. Drag a Button widget
2. Use currentItem / currentRow in binding
3. See that no lint error is shown

### Public Sample App

_No response_

### Version

Cloud

### LINKED ISSUE #15156: [Bug]: semicolon following a block or function declaration shows lint error
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

semicolon following a block or function declaration shows `Unnecessary semicolon.`, this should either be a warning or not present at all.

### Steps To Reproduce

Go to a JS enabled field and type out the following

```
{{
function example(a) {
    if (a) {
        return true;
    };
};
}}
```

### Public Sample App

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by ohansFavour**
## Description

Core changes in this PR

- Prevents logging of lint warnings (not errors) in the debugger
- Refined message for using 'await' in non-async functions
- Show lint error messages that include line numbers (previously weren't shown)

Fixes #8659
Fixes #10825
Fixes #15099 
Fixes #15156 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?
- [x] https://github.com/appsmithorg/TestSmith/issues/1938
- [x] https://github.com/appsmithorg/TestSmith/issues/1939
- [x] https://github.com/appsmithorg/TestSmith/issues/1940

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts [hunk 1]
```diff
@@ -74,15 +74,32 @@ export const getFirstNonEmptyPosition = (lines: string[]): Position => {
...
+  errors: EvaluationError[],
+  contextData?: Record<string, Record<string, unknown>>,
+) => {
+  return errors.filter(
+    (error) =>
+      error.errorType === PropertyEvaluationErrorType.LINT &&
+      // Remove all errors where additional dynamic data is reported as undefined
+      !(
+        contextData &&
+        error.code === "W117" &&
```
Comment (rimildeyjsr) [inline app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:87]: can we use a constant here, which defines what W117 is?

Comment (ApekshaBhosale) [reply app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:87]: and please add more information about that particular code error too @ohansFavour

**Comment by ramsaptami** [comment]
- [ ] Using following code does not display an error for missing semicolon
```
export default {
	newFun: () => {
		
	}
	secFun: } => {
		
	}
}
```
image


- [ ] Using following code displays error seen in the screenshot
```
export default {
	newFun: () => {
		
	}
}
	secFun: () => {
	}
}
```
image

- [ ] There's an error for missing semicolon that appears when hovering over the end of the lint error
image

**Comment by ohansFavour** [comment]
@ramsaptami all of these errors are valid and beyond our control. It's in the way JSHint analyses the code. Checkout similar behaviour [here](https://jshint.com/)

**Comment by ohansFavour** [comment]
> * [ ]  Using following code does not display an error for missing comma, instead shows a different error
> 
> ```
> export default {
> 	newFun: () => {
> 		
> 	}
> 	secFun: } => {
> 		
> 	}
> }
> ```
> 
> image
> 
> * [ ]  Using following code displays error seen in the screenshot
> 
> ```
> export default {
> 	newFun: () => {
> 		
> 	}
> }
> 	secFun: () => {
> 	}
> }
> ```
> 
> image
> 
> * [ ]  There's an error for missing semicolon that appears when hovering over the end of the lint error
>   image
> * [ ]  Incorrect error is displayed for below code
> 
> ```
> export default {
> 	func1: () => {
> 		showAlert('this')
> 	},
> 	func2: () => {
> 	}
> ```
> 
> image

@ramsaptami all of these errors are valid and beyond our control. It's in the way JSHint analyses the code. Checkout similar behaviour [here](https://jshint.com/)

File: app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts (hunk 1 above)
Comment (ohansFavour) [reply app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts]: Done

File: app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts [hunk 2]
```diff
@@ -119,7 +152,7 @@ export const getLintAnnotations = (
         if (variable) {
           variableLength =
             variableLength === 1
-              ? variable.length
+              ? String(variable).length
```
[the remark this answers is not in the document]
Comment (ohansFavour) [reply app/client/src/components/editorComponents/CodeEditor/lintHelpers.ts:155]: Sometimes, variables can be numbers (representing line numbers) and ".length" of a number is undefined. This is the reason lint errors that reference line numbers are currently not being shown.

**Review by ApekshaBhosale — APPROVED** [review]

**Review by Aishwarya-U-R — APPROVED** [review]
Approving cypress tests!

**Comment by ohansFavour** [comment]
@ramsaptami kindly add the test plan approval label after QA

**Pull request merged**